include/aikido/planner/vectorfield/BodyNodePoseVectorField.hpp

```diff
@@ -0,0 +1,124 @@
+/// used to select velocities from the null space to avoid violating
+/// joint limits.
+class BodyNodePoseVectorField : public VectorField
+{
+public:
+  /// Constructor.
+  ///
+  /// \param[in] stateSpace MetaSkeleton state space.
+  /// \param[in] bn Body node of end-effector.
+  /// \param[in] maxStepSize The maximum step size used to guarantee
+  /// that the integrator does not step out of joint limits.
+  /// \param[in] jointLimitpadding If less then this distance to joint
+  /// limit, velocity is bounded in that direction to 0.
+  BodyNodePoseVectorField(
+      aikido::statespace::dart::MetaSkeletonStateSpacePtr
+          metaSkeletonStateSpace,
+      dart::dynamics::BodyNodePtr bodyNode,
+      double maxStepSize,
+      double jointLimitPadding,
+      bool enforceJointVelocityLimits = true);
+
+  // Documentation inherited.
+  bool evaluateVelocity(
+      const aikido::statespace::StateSpace::State* state,
+      Eigen::VectorXd& qd) const override;
+
+  // Documentation inherited.
+  VectorFieldPlannerStatus evaluateStatus(
+      const aikido::statespace::StateSpace::State* state) const override;
+
+  /// Evaluate Cartesian velocity by current pose of body node.
+  ///
+  /// \param[in] pose Current pose of body node.
+  /// \param[out] cartesianVelocity Cartesian velocity defined in se(3).
+  virtual bool evaluateCartesianVelocity(
+      const Eigen::Isometry3d& pose,
+      Eigen::Vector6d& cartesianVelocity) const = 0;
+
+  /// Evaluate current pose to determine the status of planning.
+  ///
+  /// \param[in] pose Current pose of body node.
+  /// \return Status of planner.
+  virtual VectorFieldPlannerStatus evaluateCartesianStatus(
+      const Eigen::Isometry3d& pose) const = 0;
+
+  // Documentation inherited.
+  bool evaluateTrajectory(
+      const aikido::trajectory::Trajectory& trajectory,
+      const aikido::constraint::Testable* constraint,
+      double evalStepSize,
+      double& evalTimePivot,
+      bool excludeEndTime) const override;
+
+  /// Return meta skeleton state space.
+  aikido::statespace::dart::MetaSkeletonStateSpacePtr
+  getMetaSkeletonStateSpace();
+
+  /// Return const meta skeleton state space.
+  aikido::statespace::dart::ConstMetaSkeletonStateSpacePtr
+  getMetaSkeletonStateSpace() const;
+
+  /// Returns meta skeleton.
+  dart::dynamics::MetaSkeletonPtr getMetaSkeleton();
+
+  /// Returns const meta skeleton.
+  dart::dynamics::ConstMetaSkeletonPtr getMetaSkeleton() const;
+
+  /// Returns body node of end-effector.
+  dart::dynamics::BodyNodePtr getBodyNode();
+
+  /// Returns const body node of end-effector.
+  dart::dynamics::ConstBodyNodePtr getBodyNode() const;
+
+protected:
+  /// Meta skeleton state space.
+  aikido::statespace::dart::MetaSkeletonStateSpacePtr mMetaSkeletonStateSpace;
+
+  /// Meta Skeleton
+  dart::dynamics::MetaSkeletonPtr mMetaSkeleton;
+
+  /// BodyNode
+  dart::dynamics::BodyNodePtr mBodyNode;
+
+  /// Initial step size of integrator.
```

Comment:
Nit: This comment should be updated to reflect the fact that this is a maximum step size.